### PR TITLE
feat(worker): provisioned worker should not take when memory requirements

### LINKED ISSUE #3997: feat(worker): memory prerequisite and provisioned workers
provisioned workers should not take job with memory prerequisites

### PR DESCRIPTION
**Pull request by bnjjj**
1. Description
1. Related issues
close #3997 
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-cds-doc#8942.0 ✘  
* Stage 1
  * New Job ✘

File: docs/content/docs/concepts/requirement/requirement_memory.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+title: "Memory"
+weight: 6
+---
+
+The Memory requirement allow you to require a worker to have a specific number of mb of RAM.
```
Comment (yesnault) [inline docs/content/docs/concepts/requirement/requirement_memory.md]: allows

Comment (yesnault) [inline docs/content/docs/concepts/requirement/requirement_memory.md]:
```suggestion
The Memory requirement allows you to require a worker to have a specific number of mb of RAM.
```

File: engine/worker/requirement.go [hunk 2]
```diff
@@ -177,6 +158,11 @@ func checkServiceRequirement(w *currentWorker, r sdk.Requirement) (bool, error)
 }
 
 func checkMemoryRequirement(w *currentWorker, r sdk.Requirement) (bool, error) {
+	// If we are in a docker and in a provisioned worker don't run job in this worker
```
Comment (sguiheux) [inline engine/worker/requirement.go]: Is it possible for the hatchery to add metadata or env variable with the memory ?  and so use provisioned worker?

File: engine/worker/requirement.go [hunk 3]
```diff
@@ -158,24 +158,32 @@ func checkServiceRequirement(w *currentWorker, r sdk.Requirement) (bool, error)
...
-	totalMemory := v.Total
 
-	neededMemory, err := strconv.ParseInt(r.Value, 10, 64)
-	if err != nil {
-		return false, err
+	switch w.model.Type {
+	// Check env variables in a docker is safer than mem.VirtualMemory
+	case sdk.Docker:
+		var err error
+		memoryEnv := os.Getenv("CDS__MEMORY")
```
Comment (yesnault) [inline engine/worker/requirement.go]: CDS__MEMORY contains the memory = spawnArgs.Model.ModelDocker.Memory, so the memory required by the worker model.
here, we compare this memory required by the worker model with the memory of the requirement set by user.
Can you add comment about that in this func? I had to read the set of CDS__MEMORY to understand this check.

File: dist [hunk 4]
```diff
@@ -0,0 +1 @@
+/Users/bcoenen/src/github.com/ovh/cds/engine/worker/dist
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply dist]: Oupsy

File: engine/worker/requirement.go (hunk 2 above)
Comment (bnjjj) [reply engine/worker/requirement.go]: done

**Comment by ovh-cds** [comment]
CDS Report build-engine-cli-tools#9010.0 ✘  
* Build minimal
  * cdsctl linux ✔ 
  * check SQL Files number ✔ 
  * Engine Build linux amd64 ✔ 
  * tools/debpacker ✘  
  * tools/os-ansible-inventory ✔ 
  * Worker linux amd64 / 386 ✔

**Comment by ovh-cds** [comment]
CDS Report build-ui#9038.0 ✘  
* Compile
  * Build UI ✔ 
  * TU All -   
  * TU model ✔ 
  * TU service ✔ 
  * TU shared ✘  
  * TU store ✔ 
  * TU views ✘

**Pull request merged**